README.md
## Project 2 - ZETA: Network sniffer
Author: Baturov Illia (xbatur00)

## Project structure
My project has several parts. Firstly, I created global variables for each protocol, the number of packets to process, and the string to filter, as well as the device we'll be using. The important part is the argument handling, which I implemented with the function getopt_long(). This function is able to detect both long arguments and short ones. After processing the arguments, depending on the global booleans, each packet will be filtered. I use the main() function to open a session for an Ethernet device. When the session has opened successfully, you can start receiving a certain number of packets using the function pcap_loop(). As many packets will be written as specified in the global variable number_of_packets. The function pcap_loop() uses a callback function print_packet_info() that will be run for each packet. With the help of it we can get the information we need from the packet.

## Necessary theory
A packet sniffer is a tool used to analyze network traffic. It works by capturing data packets that are being transmitted over a network and then analyzing the contents of these packets. You can filter packets by different criteria, by network interface, by protocol or by port number, for example.

A network interface is a hardware component that enables a device to connect to a computer network. When a network interface receives a packet, it examines the packet's header. From the header of the packet, we can get the content of the packet, its source and destination, for example. In the sniffer, this feature is used to write out the data we need.

Also in the packet's header there is a protocol type of packet. Depending on the type of protocol, we use different headers from which we get data, different for each protocol. For example, only TCP and UDP protocols have ports.

## Code inspiration
To write a code for functions that receive packets, I was inspired by: https://www.tcpdump.org/pcap.html. This includes detecting the device, opening the session, compiling the filter, as well as looping through the received packets.

To understand how filter compilation works, I used different sites like this one: https://linux.die.net/man/7/pcap-filter. This helped me convert the text filter to the format that the sniffer needs.

It was also important for the implementation to know the number of protocols that I took from here: https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml.

## Project testing

```
sudo ./ipk-sniffer
enp0s3
any
lo
nflog

```

```
sudo ./ipk-sniffer -i enp0s3 --tcp
timestamp: 2023-04-17T21:09:16.120+0000
src MAC: 52:54:00:12:35:02
dst MAC: 08:00:27:56:aa:92
frame length: 94 bytes
src IP: 147.229.177.163
dst IP: 10.0.2.15
src port: 443
dst port: 47096

0x0000: 08 00 27 56 aa 92 52 54 00 12 35 02 08 00 45 00 ..'V..RT ..5...E.
0x0010: 00 50 ba b1 00 00 40 06 6e 5f 93 e5 b1 a3 0a 00 .P....@. n_......
0x0020: 02 0f 01 bb b7 f8 e8 b4 93 33 e7 11 56 84 50 18 ........ .3..V.P.
0x0030: ff ff 1a 16 00 00 17 03 03 00 23 3b d0 2d cd 4d ........ ..#;.-.M
0x0040: 89 14 d6 8c f6 b8 60 55 d3 f7 33 1b bb 33 92 32 ......`U ..3..3.2
0x0050: b2 0f f5 ca 21 18 4f 92 b7 28 cd f1 4d 42 00 00 ....!.O. .(..MB..

```

```
sudo ./ipk-sniffer -i enp0s3 --udp -n 3
timestamp: 2023-04-17T21:10:23.310+0000
src MAC: 08:00:27:56:aa:92
dst MAC: 52:54:00:12:35:03
frame length: 106 bytes
src IP: 10.0.2.15
dst IP: 10.0.2.3
src port: 54526
dst port: 53

0x0000: 52 54 00 12 35 03 08 00 27 56 aa 92 08 00 45 00 RT..5... 'V....E.
0x0010: 00 5c ac 05 40 00 40 11 76 7a 0a 00 02 0f 0a 00 .\..@.@. vz......
0x0020: 02 03 d4 fe 00 35 00 48 18 6b 39 91 01 00 00 01 .....5.H .k9.....
0x0030: 00 00 00 00 00 01 13 63 6f 6e 74 65 6e 74 2d 73 .......c ontent-s
0x0040: 69 67 6e 61 74 75 72 65 2d 32 03 63 64 6e 07 6d ignature -2.cdn.m
0x0050: 6f 7a 69 6c 6c 61 03 6e 65 74 00 00 01 00 01 00 ozilla.n et......
0x0060: 00 29 04 b0 00 00 00 00 00 00 00 00 00 00 00 00 .)...... ........

timestamp: 2023-04-17T21:10:23.312+0000
src MAC: 52:54:00:12:35:02
dst MAC: 08:00:27:56:aa:92
frame length: 260 bytes
src IP: 10.0.2.3
dst IP: 10.0.2.15
src port: 53
dst port: 54526

0x0000: 08 00 27 56 aa 92 52 54 00 12 35 02 08 00 45 00 ..'V..RT ..5...E.
0x0010: 00 f6 ba ba 00 00 40 11 a7 2b 0a 00 02 03 0a 00 ......@. .+......
0x0020: 02 0f 00 35 d4 fe 00 e2 b7 e3 39 91 81 80 00 01 ...5.... ..9.....
0x0030: 00 03 00 00 00 01 13 63 6f 6e 74 65 6e 74 2d 73 .......c ontent-s
0x0040: 69 67 6e 61 74 75 72 65 2d 32 03 63 64 6e 07 6d ignature -2.cdn.m
0x0050: 6f 7a 69 6c 6c 61 03 6e 65 74 00 00 01 00 01 c0 ozilla.n et......
0x0060: 0c 00 05 00 01 00 00 00 5e 00 3a 18 63 6f 6e 74 ........ ^.:.cont
0x0070: 65 6e 74 2d 73 69 67 6e 61 74 75 72 65 2d 63 68 ent-sign ature-ch
0x0080: 61 69 6e 73 04 70 72 6f 64 09 61 75 74 6f 67 72 ains.pro d.autogr
0x0090: 61 70 68 08 73 65 72 76 69 63 65 73 06 6d 6f 7a aph.serv ices.moz
0x00a0: 61 77 73 c0 2c c0 41 00 05 00 01 00 00 00 5e 00 aws.,.A. ......^.
0x00b0: 38 04 70 72 6f 64 18 63 6f 6e 74 65 6e 74 2d 73 8.prod.c ontent-s
0x00c0: 69 67 6e 61 74 75 72 65 2d 63 68 61 69 6e 73 04 ignature -chains.
0x00d0: 70 72 6f 64 0b 77 65 62 73 65 72 76 69 63 65 73 prod.web services
0x00e0: 06 6d 6f 7a 67 63 70 c0 2c c0 87 00 01 00 01 00 .mozgcp. ,.......
0x00f0: 00 00 5e 00 04 22 a0 90 bf 00 00 29 ff d6 00 00 ..^..".. ...)....
0x0100: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ........ ........

timestamp: 2023-04-17T21:10:23.312+0000
src MAC: 08:00:27:56:aa:92
dst MAC: 52:54:00:12:35:03
frame length: 106 bytes
src IP: 10.0.2.15
dst IP: 10.0.2.3
src port: 54526
dst port: 53

0x0000: 52 54 00 12 35 03 08 00 27 56 aa 92 08 00 45 00 RT..5... 'V....E.
0x0010: 00 5c ac 06 40 00 40 11 76 79 0a 00 02 0f 0a 00 .\..@.@. vy......
0x0020: 02 03 d4 fe 00 35 00 48 18 6b f3 96 01 00 00 01 .....5.H .k......
0x0030: 00 00 00 00 00 01 13 63 6f 6e 74 65 6e 74 2d 73 .......c ontent-s
0x0040: 69 67 6e 61 74 75 72 65 2d 32 03 63 64 6e 07 6d ignature -2.cdn.m
0x0050: 6f 7a 69 6c 6c 61 03 6e 65 74 00 00 1c 00 01 00 ozilla.n et......
0x0060: 00 29 04 b0 00 00 00 00 00 00 00 00 00 00 00 00 .)...... ........

```

## Bibliography
* http://www.tcpdump.org/
* http://en.wikipedia.org/wiki/Pcap
* https://www.geeksforgeeks.org/introduction-to-sniffers/


CHANGELOG.md
- Changed timestamp function and minor changes
- Minor changes
- Added IPv6 and changed filter
- Added ICMP and UDP protocols to packet handler function
- Added working Makefile
- Added print packet content function
- Partially added IPv4 packet information
- Minor changes and added signal handler
- Added src MAC and dst MAC, changed the number of packets to 1
- Added packet looping function and print packet information function
- Added packet filter basis, simple tcp protocol filtering
- Created basis for sniffering, takes only one packet and writes out its length
- Minor changes, added parse_args() function
- Added print_timestamp() function that outputs current time in the required format
- Added print_interfaces() function that outputs all available interfaces
- Added argument parsing with getopt_long() function
- Added initialization of all values and no parameters test
- Added print usage function, changed libraries
- Initial commit


ipk-sniffer.cpp
// VUT FIT - IPK
// Project 2 - ZETA: Network sniffer
// Author: Baturov Illia (xbatur00)

// Libraries
#include <iostream>
#include <string>
#include <cstring>
#include <sstream>
#include <getopt.h>
#include <pcap/pcap.h>
#include <ctime>
#include <iomanip>
#include <signal.h>
#include <net/ethernet.h>
#include <netinet/ip.h>
#include <netinet/ip6.h>
#include <netinet/tcp.h>
#include <netinet/udp.h>
#include <netinet/in.h>
#include <arpa/inet.h>

using namespace std;

// Global variables
string interface;
bool tcp, udp, icmp4, icmp6, arp, ndp, igmp, mld = false;
int port = -1;
int number_of_packets = 1;
string filter = ""; // The filter expression
struct bpf_program fp; // The compiled filter
pcap_t *handle; // Session handle

// Usage
void print_usage() {
    cout << "Usage: ipk-sniffer [-i interface | --interface interface] {-p port [--tcp|-t] [--udp|-u]} [--arp] [--icmp4] [--icmp6] [--igmp] [--mld] {-n num}" << endl;
    return;
}

// Print all interfaces
void print_interfaces() {
    char errbuf[PCAP_ERRBUF_SIZE];
    pcap_if_t *all_interfaces;
    if (pcap_findalldevs(&all_interfaces, errbuf) == -1) {
        cerr << "ERROR: No devices. INFO: " << errbuf;
        exit(EXIT_FAILURE);
    }
    while(all_interfaces->next != NULL) {
        cout << all_interfaces->name << endl;
        all_interfaces = all_interfaces->next;
    }
    pcap_freealldevs(all_interfaces);
    exit(0);
}

// Print timestamp in RFC 3339 format
void print_timestamp(const struct pcap_pkthdr *header) {
    // Get the current timestamp
    time_t s = header->ts.tv_sec;
    suseconds_t ms = header->ts.tv_usec;

    // Convert to the required format
    char timestamp[30];
    strftime(timestamp, sizeof(timestamp), "%Y-%m-%dT%H:%M:%S", localtime(&s));
    sprintf(timestamp + strlen(timestamp), ".%03ld", ms / 1000);
    // Note: Dividing by 1000 to remove the seconds.
    strftime(timestamp + strlen(timestamp), sizeof(timestamp) - strlen(timestamp), "%z", localtime(&s));

    // Print the timestamp
    cout << "timestamp: " << timestamp << endl;
    // Format: 2021-03-19T18:42:52.362+01:00
    return;
}

// Parse program arguments
void parse_arguments(int argc, const char **argv) {
    // If no parameters are specified, print a list of active interfaces
    if (argc == 1) {
        print_interfaces();
        exit(0);
    }

    // Array of struct option structures needed for getopt_long() function
    static struct option options[] = {
        {"interface", required_argument, 0, 'i'},
        {"tcp", no_argument, 0, 't'},
        {"udp", no_argument, 0, 'u'},
        {"icmp4", no_argument, 0, 0},
        {"icmp6", no_argument, 0, 0},
        {"arp", no_argument, 0, 0},
        {"ndp", no_argument, 0, 0},
        {"igmp", no_argument, 0, 0},
        {"mld", no_argument, 0, 0},
        {0, 0, 0, 0}
    };

    // Parse all arguments
    char** args = (char**)argv;
    // Note: Cast to avoid getopt_long() error, argv is const char**, function need const* char* parameter.
    int index = 0;
    int option;
    while ((option = getopt_long(argc, args, "i:p:tun:", options, &index)) != -1) {
        switch(option) {
            case 'i':
                interface = optarg;
                break;
            case 'p':
                port = stoi(optarg);
                break;
            case 't':
                tcp = true;
                break;
            case 'u':
                udp = true;
                break;
            case 'n':
                number_of_packets = stoi(optarg);
                break;
            case 0:
                if (string(options[index].name) == "icmp4") {
                    icmp4 = true;
                } else if (string(options[index].name) == "icmp6") {
                    icmp6 = true;
                } else if (string(options[index].name) == "arp") {
                    arp = true;
                } else if (string(options[index].name) == "ndp") {
                    ndp = true;
                } else if (string(options[index].name) == "igmp") {
                    igmp = true;
                } else if (string(options[index].name) == "mld") {
                    mld = true;
                }
                break;
            default:
                cerr << "ERROR: Wrong program arguments." << endl;
                print_usage();
                exit(EXIT_FAILURE);
        }
    }
}

// Create a protocol filter
void create_filter() {
    // 
    if (port == -1) {
        if (tcp) {
            filter += "tcp or ";
        }
        if (udp) {
            filter += "udp or ";
        }
    } else {
        if (tcp) {
            filter += "tcp and port " + to_string(port) + " or ";
        }
        if (udp) {
            filter += "udp and port " + to_string(port) + " or ";
        }
    }

    // Add protocol filter
    if (icmp4) {
        filter += "icmp or ";
    }
    if (icmp6) {
        filter += "icmp6 or ";
    }
    if (arp) {
        filter += "arp or ";
    }
    if (ndp) {
        filter += "icmp6 and icmp6[0] = 135 or ";
    }
    if (igmp) {
        filter += "igmp or ";
    }
    if (mld) {
        filter += "icmp6 and icmp6[0] = 130 or ";
    }

    // Delete last 3 characters
    if (filter != "") {
        filter.erase(filter.size() - 3);
    }

    return;
}

// Print byte_offset_hexa byte_offset_ASCII
void print_packet_content(const u_char* packet, int length) {
    int zero = 0;
    int byte_count = 0;
    int byte_offset = 0;
    const int bytes_per_row = 16;

    // Loop through the packet printing out each byte
    cout << endl;
    while (byte_offset < length) {
        // Print the byte offset in hexadecimal format
        printf("0x%04x: ", byte_offset);

        // Print out the bytes in hexadecimal format
        for (int i = 0; i < bytes_per_row; i++) {
            if (byte_offset < length) {
                printf("%02x ", packet[byte_offset]);
            } else {
                printf("%02x ", zero);
            }
            byte_offset++;
            byte_count++;
        }

        // Print out the ASCII representation of the bytes
        for (int i = byte_count - bytes_per_row; i < byte_count; i++) {
            if (byte_count - i == 8) {
                printf("%c", ' ');
            }
            // Note: Print white space after 8th byte.
            if (i < length) {
                if (isprint(packet[i])) {
                    printf("%c", packet[i]);

                } else {
                    printf("%c",'.');
                }
            } else {
                printf("%c",'.');
            }
        }
        cout << endl;
    }
    return;
}

// Callback function for looping through packets
void print_packet_info(unsigned char *args, const struct pcap_pkthdr *header, const unsigned char *packet) {
    // Variables
    const struct ether_header *ether_h = (const struct ether_header *)packet; // ethernet header
    const struct ip *ipv4_h = (const struct ip *)(packet + sizeof(struct ether_header)); // ipv4 header
    const struct ip6_hdr* ipv6_h = (struct ip6_hdr*)(packet + sizeof(struct ether_header)); // ipv6 header
    const struct tcphdr *tcp_h = (struct tcphdr *)(packet + sizeof(struct ether_header) + sizeof(struct ip)); // tcp header
    const struct udphdr *udp_h = (struct udphdr *)(packet + sizeof(struct ether_header) + sizeof(struct ip)); // udp header

    print_timestamp(header); // timestamp
    printf("src MAC: %02x:%02x:%02x:%02x:%02x:%02x\n", \
            ether_h->ether_shost[0], ether_h->ether_shost[1], ether_h->ether_shost[2], \
            ether_h->ether_shost[3], ether_h->ether_shost[4], ether_h->ether_shost[5]); // src MAC
    printf("dst MAC: %02x:%02x:%02x:%02x:%02x:%02x\n", \
            ether_h->ether_dhost[0], ether_h->ether_dhost[1], ether_h->ether_dhost[2], \
            ether_h->ether_dhost[3], ether_h->ether_dhost[4], ether_h->ether_dhost[5]); // dst MAC
    printf("frame length: %d bytes\n", header->len); // frame length

    // IPv4
    if (ntohs(ether_h->ether_type) == ETHERTYPE_IP) {
        // ICMP protocol
        if (ipv4_h->ip_p == 1) {
            // Print the packet information
            printf("src IP: %s\n", inet_ntoa(ipv4_h->ip_src)); // src IP
            printf("dst IP: %s\n", inet_ntoa(ipv4_h->ip_dst)); // dst IP
        // TCP protocol
        } else if (ipv4_h->ip_p == 6) {
            // Print the packet information
            printf("src IP: %s\n", inet_ntoa(ipv4_h->ip_src)); // src IP
            printf("dst IP: %s\n", inet_ntoa(ipv4_h->ip_dst)); // dst IP
            printf("src port: %d\n", ntohs(tcp_h->th_sport)); // src port
            printf("dst port: %d\n", ntohs(tcp_h->th_dport)); // dst port
        // UDP protocol
        } else if (ipv4_h->ip_p == 17) {
            // Print the packet information
            printf("src IP: %s\n", inet_ntoa(ipv4_h->ip_src)); // src IP
            printf("dst IP: %s\n", inet_ntoa(ipv4_h->ip_dst)); // dst IP
            printf("src port: %d\n", ntohs(udp_h->uh_sport)); // src port
            printf("dst port: %d\n", ntohs(udp_h->uh_dport)); // dst port
        }
    // IPv6
    } else if (ntohs(ether_h->ether_type) == ETHERTYPE_IPV6) {
        char src_ipv6[INET6_ADDRSTRLEN], dst_ipv6[INET6_ADDRSTRLEN];
        // Note: Convert IPv6 address to string.
        inet_ntop(AF_INET6, &(ipv6_h->ip6_src), src_ipv6, INET6_ADDRSTRLEN);
        inet_ntop(AF_INET6, &(ipv6_h->ip6_dst), dst_ipv6, INET6_ADDRSTRLEN);
        // TCP protocol
        if (ipv6_h->ip6_nxt == 6) {
            // Print the packet information
            printf("src IP: %s\n", src_ipv6); // src IP
            printf("dst IP: %s\n", dst_ipv6); // dst IP
            printf("src port: %d\n", ntohs(tcp_h->th_sport)); // src port
            printf("dst port: %d\n", ntohs(tcp_h->th_dport)); // dst port
        // UDP protocol
        } else if (ipv6_h->ip6_nxt == 17) {
            // Print the packet information
            printf("src IP: %s\n", src_ipv6); // src IP
            printf("dst IP: %s\n", dst_ipv6); // dst IP
            printf("src port: %d\n", ntohs(udp_h->uh_sport)); // src port
            printf("dst port: %d\n", ntohs(udp_h->uh_dport)); // dst port
        // ICMP protocol
        } else if (ipv6_h->ip6_nxt == 58) {
            // Print the packet information
            printf("src IP: %s\n", src_ipv6); // src IP
            printf("dst IP: %s\n", dst_ipv6); // dst IP
        }
    }
    print_packet_content(packet, header->len); // print packet content
    cout << endl;
    return;
}

// Function for CTRL + C
void handle_signal(int signum)
{
    cout << endl << "Received signal: " << signum << ". Exiting..." << endl;
    pcap_close(handle);
    exit(signum);
}

// Main
int main (int argc, const char *argv[]) {
    char errbuf[PCAP_ERRBUF_SIZE]; // Error string
    bpf_u_int32 net; // IP address of device
    bpf_u_int32 mask; // Netmask of device
    struct pcap_pkthdr header; // pcup packet header
    const u_char *packet; // actual packet

    signal(SIGINT, handle_signal); // Handle a signal

    parse_arguments(argc, argv); // Parse arguments

    create_filter(); // Create a filter

    // Find the properties for the device
    if (pcap_lookupnet(interface.c_str(), &net, &mask, errbuf) == PCAP_ERROR) {
        cerr << "ERROR: Can't get netmask of device. INFO: " << errbuf;
        exit(EXIT_FAILURE);
    }
    // Open the session
    handle = pcap_open_live(interface.c_str(), BUFSIZ, 1, 1000, errbuf);
    if (handle == NULL) {
        cerr << "ERROR: Can't open device. INFO: " << errbuf;
        exit(EXIT_FAILURE);
    }
    // Check if device is an Ethernet device
    if (pcap_datalink(handle) != DLT_EN10MB) {
        // Note: Data link type of an Ethernet devise is "DLT_EN10MB".
        pcap_close(handle);
        cerr << "ERROR: Device is not an Ethernet device." << errbuf;
        exit(EXIT_FAILURE);
    }
    // Filtering
    if (filter != "") {
        // Compile the filter
        if (pcap_compile(handle, &fp, filter.c_str(), 0, net) == -1) {
            pcap_close(handle);
            cerr << "ERROR: Can't compile filter. INFO: " << pcap_geterr(handle);
            exit(EXIT_FAILURE);
        }
        // Apply the filter
        if (pcap_setfilter(handle, &fp) == -1) {
            pcap_close(handle);
            cerr << "ERROR: Can't apply filter. INFO: " << pcap_geterr(handle);
            exit(EXIT_FAILURE);
        }
    }
    // Grab a packet
    if (pcap_loop(handle, number_of_packets, print_packet_info, nullptr) == PCAP_ERROR) {
        pcap_close(handle);
        cerr << "ERROR: Can't read packets. INFO: " << pcap_geterr(handle);
        exit(EXIT_FAILURE);
    }

    // Close the session
    pcap_close(handle);
    return 0;
}

// End of program
